suskin/stack-template-engine
# Add resource engine interface

## Overview

The render phase controller was becoming overly large and a bit hard to read, so it was about time to refactor it a little bit. Also, @muvaf's upcoming work on adding another engine type means that it'd be helpful to have a better contract for decoupling the resource engine logic from the controller flow.

This change aims to do two things:

* Introduce a new, cleaner contract for resource engines (this is also the first pass at the contract)
* Move resource engine logic out of the render controller so that it isn't cluttering things up for readers and maintainers

The commits in the PR should tell a relatively reasonable story about how this happened. But, to summarize, most of the changes are refactoring and moving the existing logic behind the interface. There's also a minor change to inject the name of the render controller's stack configuration in at the time it's created; previously it was being assumed to be the same as the claim that was being processed, which isn't going to work for real use-cases.

## Testing done

I've been testing this locally similar to the previous PRs. The gist of it is:

* Create a StackConfiguration by hand
* Create an object mentioned in that StackConfiguration (I've been using `HelmChartInstall`)
* `make run`
* Check that the jobs ran successfully and objects were created

Pretty soon I plan to investigate the [integration testing offered by controller-runtime](https://godoc.org/sigs.k8s.io/controller-runtime#hdr-Testing), but I haven't quite gotten around to it yet.

## Future work

As was noted in an earlier PR, there is some hardening that will be needed for the render controller. None of it is included in this PR.

* [ ] Only create one render controller per CRD. Currently one per CRD is created for every Reconcile call in the setup controller.
* [ ] Stop render controllers when they're no longer needed.
* [ ] Only create resources when it's needed, rather than blindly on every reconcile.
